upgradeinterlockccl: failure in TestTenantUpgradeInterlock

On an unrelated [PR](https://teamcity.cockroachdb.com/viewLog.html?buildId=11349962&buildTypeId=Cockroach_BazelEssentialCi) on 7da4ef5af83c38680c536e7c18a9db61274b9d25

```
=== RUN   TestTenantUpgradeInterlock_current_binary_version_pause_after_migration
I230816 18:30:28.272025 1 (gostd) rand.go:249  [-] 1  random seed: -1349245253279896280
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/7dbbbba0c50ac3c168da74606a55b80c/logTestTenantUpgradeInterlock_current_binary_version_pause_after_migration421530839
    test_log_scope.go:81: use -show-logs to present logs inline
    local_test_util_test.go:51: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) upgrade interlock test: running variant "current_binary_version", configuration: "pause_after_migration"
    local_test_util_test.go:167: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) creating an initial tenant server
    local_test_util_test.go:175: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) verifying the tenant version
    local_test_util_test.go:179: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) verifying basic SQL functionality
    local_test_util_test.go:184: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) verifying the version of the storage cluster
    local_test_util_test.go:191: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) upgrading the storage cluster
    local_test_util_test.go:194: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) checking the tenant after the storage cluster upgrade
    local_test_util_test.go:198: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) start upgrading the tenant
    local_test_util_test.go:253: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) upgrader is ready
    local_test_util_test.go:255: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) starting another tenant server
    local_test_util_test.go:301: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) waiting for the instance table to get in the right state
    local_test_util_test.go:312: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) resuming upgrade
    local_test_util_test.go:314: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) waiting for upgrade to complete
    local_test_util_test.go:245: unexpected error:  pq: upgrade failed due to transient SQL servers
    local_test_util_test.go:316: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) upgrade completed
    local_test_util_test.go:335: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) check the second server still works
    local_test_util_test.go:337: (TestTenantUpgradeInterlock_current_binary_version_pause_after_migration) waiting for second server to reach target final version
    sql_runner.go:104: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:117
        	            				github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:312
        	            				pkg/ccl/kvccl/kvtenantccl/upgradeinterlockccl/upgradeinterlockccl_test/pkg/ccl/kvccl/kvtenantccl/upgradeinterlockccl/local_test_util_test.go:338
        	            				pkg/ccl/kvccl/kvtenantccl/upgradeinterlockccl/upgradeinterlockccl_test/bazel-out/k8-fastbuild/bin/pkg/ccl/kvccl/kvtenantccl/upgradeinterlockccl/generated_test.go:96
        	Error:      	Received unexpected error:
        	            	query 'SHOW CLUSTER SETTING version': expected:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:315
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:75
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:81
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:117
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:312
        	            	  | pkg/ccl/kvccl/kvtenantccl/upgradeinterlockccl/upgradeinterlockccl_test.runTest
        	            	  | 	pkg/ccl/kvccl/kvtenantccl/upgradeinterlockccl/upgradeinterlockccl_test/pkg/ccl/kvccl/kvtenantccl/upgradeinterlockccl/local_test_util_test.go:338
        	            	  | pkg/ccl/kvccl/kvtenantccl/upgradeinterlockccl/upgradeinterlockccl_test.TestTenantUpgradeInterlock_current_binary_version_pause_after_migration
        	            	  | 	pkg/ccl/kvccl/kvtenantccl/upgradeinterlockccl/upgradeinterlockccl_test/bazel-out/k8-fastbuild/bin/pkg/ccl/kvccl/kvtenantccl/upgradeinterlockccl/generated_test.go:96
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) query 'SHOW CLUSTER SETTING version': expected:
        	            	  | 1000023.1-18
        	            	  |
        	            	  | got:
        	            	  | 1000022.2-1
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/7dbbbba0c50ac3c168da74606a55b80c/logTestTenantUpgradeInterlock_current_binary_version_pause_after_migration421530839
--- FAIL: TestTenantUpgradeInterlock_current_binary_version_pause_after_migration (96.54s)
```

Jira issue: CRDB-30669

--- a/pkg/upgrade/upgradecluster/tenant_cluster.go
+++ b/pkg/upgrade/upgradecluster/tenant_cluster.go
@@ -29,6 +29,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/redact"
+	"google.golang.org/grpc"
 )
 
 // TenantCluster represents the set of sql nodes running in a secondary tenant.
@@ -228,8 +229,19 @@ func (t *TenantCluster) ForEveryNodeOrServer(
 		grp.GoCtx(func(ctx context.Context) error {
 			defer alloc.Release()
 
-			conn, err := t.Dialer.Dial(ctx, roachpb.NodeID(instance.InstanceID), rpc.DefaultClass)
-			if err != nil {
+			var conn *grpc.ClientConn
+			retryOpts := retry.Options{
+				InitialBackoff: 0,
+				MaxRetries:     2,
+				MaxBackoff:     10 * time.Millisecond,
+			}
+			// This retry was added to benefit our tests (not users) by reducing the chance of
+			// test flakes due to network issues.
+			if err := retry.WithMaxAttempts(ctx, retryOpts, retryOpts.MaxRetries+1, func() error {
+				var err error
+				conn, err = t.Dialer.Dial(ctx, roachpb.NodeID(instance.InstanceID), rpc.DefaultClass)
+				return err
+			}); err != nil {
 				if errors.HasType(err, (*netutil.InitialHeartbeatFailedError)(nil)) {
 					if errors.Is(err, rpc.VersionCompatError) {
 						return errors.WithHint(errors.Newf("upgrade failed due to active SQL servers with incompatible binary version(s)"),

